src/coreclr/tools/Common/TypeSystem/IL/UnsafeAccessors.cs

```diff
@@ -366,10 +367,22 @@ private static bool DoesMethodMatchUnsafeAccessorDeclaration(ref GenerationConte
             return true;
         }
 
-        private static bool TrySetTargetMethod(ref GenerationContext context, string name, out bool isAmbiguous, bool ignoreCustomModifiers = true)
+        private static unsafe bool TrySetTargetMethod(ref GenerationContext context, string name, out bool isAmbiguous, bool ignoreCustomModifiers = true)
         {
             TypeDesc targetType = context.TargetType;
 
+            // Build up a substitution Instantiation to use when looking up methods involving generics.
+            Instantiation substitutionForSig = default;
+            if (targetType.Instantiation.Length > 0)
+            {
+                TypeDesc[] types = new TypeDesc[targetType.Instantiation.Length];
+                for (int i = 0; i < types.Length; ++i)
+                {
+                    types[i] = targetType.Context.GetSignatureVariable(i, true);
```

Comment:
Are we confusing type generic parameters with method generic parameters here? (I think this should be passing `false` not, `true`.)

But is this substitution really needed? It looks like a no-op to me (it will replace `!0`, `!1`... with a `!0`, `!1`...).

Reply:
Hmmm. Something is amiss here. I agree this is wrong, but now I'm not sure how this is passing now.

Reply:
I think this is correct actually. Passing true indicates the type is a method. So I am substituting types for methods in the call below. I've also confirmed that is how it looks when stepping through. Am I miss understanding how this API works?

Reply:
Here's what's broken:

```csharp
using System;
using System.Runtime.CompilerServices;

class Program
{
    static void Main()
    {
        Acc<object>.M<string>(null, null, null);
    }

    class Acc<T>
    {
        [UnsafeAccessor(UnsafeAccessorKind.StaticMethod, Name = "Method")]
        public static extern void M<U>(Gen<T> owner, T t, U u);
    }
}

class Gen<T>
{
    public static void Method<U>(T t, U u) => Console.WriteLine("Hello");
}
```

The signature we're looking for is `(!0, !!0)`. This logic changes it to `(!!0, !!0)` (because we're replacing VAR with MVAR due to the `true`) and then we don't find the method with such signature.

If I delete the entire `substitutionForSig` part, the above repro "just works".

Reply:
Yes, thanks. I spoke to David yesterday and he walked me through my confusion. Appreciate the concrete example as that confirms the test I wrote wasn't updated properly so was "working".